src/inmanta/plugins.py

```diff
@@ -544,3 +564,26 @@ def wrapper(self, *args: object, **kwargs: object) -> Any:
     elif function is not None:
         fnc = curry_name(commands=commands, emits_statements=emits_statements, allow_unknown=allow_unknown)
         return fnc(function)
+
+
+@stable_api
+def deprecated(
+    function: Optional[Callable] = None, *, replaced_by: Optional[str] = None, **kwargs: abc.Mapping[str, object]
+) -> Callable:
+    """
+    the kwargs are currently ignored but where added in case we want to add something later on.
+    """
+
+    def inner(fnc: Callable):
+        if hasattr(fnc, "__plugin__"):
+            fnc.__plugin__.deprecate_function(replaced_by)
+        else:
+            raise Exception(
+                f"Can not deprecate '{fnc.__name__}': The '@deprecated' decorator should be used in combination with the "
+                f"'@plugin' decorator and should be placed first."
```

Comment:
Only noticed this on my nth read of this line but "first" is actually quite ambiguous. It could mean first as we intend or as the inner decorator (the first wrapper). I'm not sure if my suggestion is good.
```suggestion
                f"'@plugin' decorator and should be placed at the top."
```